Protocol redefinition case and defmulti case

ericnormand opened the pull request:
Redefining a protocol loses all of the types that abide by it. If you redefine the protocol, which could be in file A, you need to rerun all `deftypes`, `defrecords`, `extend-types`, and `extend-protocols` that refer to it, which could be in any number of files B, C, ..., Z.

Re-evaluating a `defmulti` has no effect if the name is unchanged. If you redefine the dispatch function, it will not change the original multi, and will thus give you incorrect results compared to removing the `defmulti`'s Var and compiling it again.

ericnormand committed:
Add case of redefining a protocol affects types implementing it

ericnormand committed:
Add multimethod redefinition test.

Defining a multimethod a second time has no effect.

aredington commented on src/clojure_repl_sufficient_p/defmulti.clj:
```diff
@@ -0,0 +1,22 @@
+
+(defmulti color-change (fn [x & _] (:or (:animal x)
+                                        (:plant  x))))
+
```
For this to be a good case, we should have a `(defmethod color-change :hydrangea ...)` case at the minimum given the expectations of the test. 

As it turns out, I scrapped a multimethod namespace because it was surprising me with how well it behaved. Which is not to say we can't find edge cases, but we do need to be crisp about them actually being Clojure's problems. :)

ericnormand replied:
Agreed. It was an oversight.

aredington reviewed (changes requested):
I think exploring more of the warts around defprotocol is a great idea. Let's attack the entire space:

- defrecord implicitly implementing a protocol, and the protocol is redefined
- deftype implicitly implementing a protocol, and the protocol is redefined

- defrecord implicitly implementing a protocol, and the record is redefined
- deftype implicitly implementing a protocol, and the type is redefined

- extend-protocol on extant record and protocol, redefine protocol
- extend-protocol on extant record and protocol, redefine record
- extend-protocol on extant type and protocol, redefine protocol
- extend-protocol on extant type and protocol, redefine type
- extend-type on same 4-branches cartesian product as prior

A lot of these things WORK. That's good, we want our programming language to work. Discussing all of this before merging isn't to dump a load of work in your lap, let's divide and conquer up this description of work. But let's do something more organized than `defprotocol.clj` `defprotocol2.clj` ... etc

ericnormand committed:
Improve defmulti tests

The iteration case was lacking a defmethod for the
:hydrangea dispatch value. It was also using a funky caught
error approach which was unnecessary. Instead, we implement
the default dispatch value case with a sentinal value.

The deftest also does not have to dump the values into
hashmaps, since it is not working with records.

I've also made the test clean up the namespace before and
after, and force a namespace reload, because this code
is (at the moment) not reloadable in the normal way.

ericnormand committed:
Define 14 scenarios for evaling protocol expressions

ericnormand commented:
Okay, I've covered 14 scenarios dealing with protocols.

aredington commented:
I'm going to add some editorial notes and also some passing test cases. Particularly because changing the default of a defmulti DOES work correctly, so changing the dispatch fn and it not working highlights how this particular case really causes bad interface consistency. (That is, the interface Clojure offers its developer users.)
